working/1426-extension-types/feature-specification-views.md

```diff
@@ -0,0 +1,955 @@
+the view declaration, the scoping and the type and meaning of `this` is the
+same as for extension methods (a feature which was added to Dart in version
+2.6). This is important because it implies that the language Dart has a
+single and consistent semantics for all statically resolved member
+invocations, rather than having one set of rules for extension methods, and
+a different set of rules for view members.
+
+
+## Motivation
+
+A _view_ is a zero-cost abstraction mechanism that allows
+developers to replace the set of available operations on a given object
+(that is, the instance members of its type) by a different set of
+operations (the members declared by the given view type).
+
+It is zero-cost in the sense that the value denoted by an expression whose
+type is a view type is an object of a different type (known as the
+on-type of the view type), there is no wrapper object.
+
+The point is that the view type allows for a convenient and safe treatment
+of a given object `o` (and objects reachable from `o`) for a specialized
+purpose. It is in particular aimed at the situation where that purpose
+requires a certain discipline in the use of `o`'s instance methods: We may
+call certain methods, but only in specific ways, and other methods should
+not be called at all. This kind of added discipline can be enforced by
+accessing `o` typed as a view type, rather than typed as its run-time
+type `R` or some supertype of `R` (which is what we normally do).
+
+A potential application would be generated view declarations handling the
+navigation of dynamic object trees. For instance, they could be JSON
+values, modeled using `num`, `bool`, `String`, `List<dynamic>`, and
+`Map<String, dynamic>`.
+
+Without view types, the JSON value would most likely be handled with
+static type `dynamic`, and all operations on it would be unsafe. If the
+JSON value is assumed to satisfy a specific schema, then it would be
+possible to reason about this dynamic code and navigate the tree correctly
+according to the schema. However, the code where this kind of careful
+reasoning is required may be fragmented into many different locations, and
+there is no help detecting that some of those locations are treating the
+tree incorrectly according to the schema.
+
+If views are supported, we can declare a set of view types with
+operations that are tailored to work correctly with the given schema and
+its subschemas. This is less error-prone and more maintainable than the
+approach where the tree is handled with static type `dynamic` everywhere.
+
+Here's an example that shows the core of that scenario. The schema that
+we're assuming allows for nested `List<dynamic>` with numbers at the
+leaves, and nothing else.
+
+```dart
+view TinyJson on Object {
+  Iterable<num> get leaves sync* {
+    var self = this;
+    if (self is num) {
+      yield self;
+    } else if (self is List<dynamic>) {
+      for (var element in self) {
+        yield* element.leaves;
```

Comment:
I think this code is quadratic, since every time you yield an element, you re-traverse element.leaves.  Might be better to use a different example, or not use sync*?

Reply:
The number of invocations of `leaves` is actually identical to the number of nodes in the given tree, so if there is anything which is too expensive it would be the use of `sync*` (it is known that `sync*` ought to be optimized better than it is today).

I ran the following:

```dart
class TinyJson {
  final Object self;

  TinyJson(this.self);

  Iterable<num> get leaves sync* {
    var self = this.self;
    if (self is num) {
      yield self;
    } else if (self is List<dynamic>) {
      for (var element in self) {
        yield* TinyJson(element).leaves;
      }
    } else {
      throw "Unexpected object encountered in TinyJson value";
    }
  }
}

var b = false;

void use(dynamic d) {
  if (b) d.fooBar(); // Pretend that `d` is not ignored.
  try {} finally {} // Prevent inlining.
}

void testFlatList() {
  print('---------- FlatList ----------');
  for (int length = 1024; length < 1000000000; length *= 2) {
    var tiny = TinyJson(List<dynamic>.filled(length, 1));
    var sw = Stopwatch()..start();
    use(tiny.leaves);
    print('Length: $length, time: ${sw.elapsed}');
  }
}

void testTwoLevelList() {
  print('---------- TwoLevelList ----------');
  for (int length = 1024; length < 1000000000; length *= 2) {
    var tiny = TinyJson(List<dynamic>.filled(length, <dynamic>[1, 2]));
    var sw = Stopwatch()..start();
    use(tiny.leaves);
    print('Length: $length, time: ${sw.elapsed}');
  }
}

void testManyLevelList() {
  print('---------- TwoLevelList ----------');
  var max = 1000;
  List<dynamic> tinies = List.filled(max, [<dynamic>[0]]);
  for (int depth = 1; depth < max; depth++) {
    var tiny = tinies[depth] = <dynamic>[tinies[depth - 1]];
    var sw = Stopwatch()..start();
    use(TinyJson(tiny).leaves);
    print('Depth: $depth, time: ${sw.elapsed}, tiny: $tiny');
  }
}

void main() {
  testFlatList();
  testTwoLevelList();
  testManyLevelList();
}
```

I got the following output, which doesn't seem to imply anything excessively slow:

```
---------- FlatList ----------
Length: 1024, time: 0:00:00.001751
Length: 2048, time: 0:00:00.000004
Length: 4096, time: 0:00:00.000001
Length: 8192, time: 0:00:00.000001
Length: 16384, time: 0:00:00.000001
Length: 32768, time: 0:00:00.000001
Length: 65536, time: 0:00:00.000003
Length: 131072, time: 0:00:00.000005
Length: 262144, time: 0:00:00.000002
Length: 524288, time: 0:00:00.000002
Length: 1048576, time: 0:00:00.000002
Length: 2097152, time: 0:00:00.000002
Length: 4194304, time: 0:00:00.000008
Length: 8388608, time: 0:00:00.000008
Length: 16777216, time: 0:00:00.000006
Length: 33554432, time: 0:00:00.000007
Length: 67108864, time: 0:00:00.000008
Length: 134217728, time: 0:00:00.000008
Length: 268435456, time: 0:00:00.000007
Length: 536870912, time: 0:00:00.000008
---------- TwoLevelList ----------
Length: 1024, time: 0:00:00.000006
Length: 2048, time: 0:00:00.000000
Length: 4096, time: 0:00:00.000000
Length: 8192, time: 0:00:00.000000
Length: 16384, time: 0:00:00.000000
Length: 32768, time: 0:00:00.000008
Length: 65536, time: 0:00:00.000003
Length: 131072, time: 0:00:00.000003
Length: 262144, time: 0:00:00.000003
Length: 524288, time: 0:00:00.000002
Length: 1048576, time: 0:00:00.000001
Length: 2097152, time: 0:00:00.000001
Length: 4194304, time: 0:00:00.000002
Length: 8388608, time: 0:00:00.000004
Length: 16777216, time: 0:00:00.000007
Length: 33554432, time: 0:00:00.000007
Length: 67108864, time: 0:00:00.000007
Length: 134217728, time: 0:00:00.000007
Length: 268435456, time: 0:00:00.000007
Length: 536870912, time: 0:00:00.000008
---------- TwoLevelList ----------
Depth: 1, time: 0:00:00.000008
Depth: 2, time: 0:00:00.000000
Depth: 3, time: 0:00:00.000000
Depth: 4, time: 0:00:00.000000
Depth: 5, time: 0:00:00.000000
Depth: 6, time: 0:00:00.000000
Depth: 7, time: 0:00:00.000000
Depth: 8, time: 0:00:00.000000
Depth: 9, time: 0:00:00.000000
Depth: 10, time: 0:00:00.000000
...
Depth: 990, time: 0:00:00.000000
Depth: 991, time: 0:00:00.000000
Depth: 992, time: 0:00:00.000000
Depth: 993, time: 0:00:00.000000
Depth: 994, time: 0:00:00.000000
Depth: 995, time: 0:00:00.000000
Depth: 996, time: 0:00:00.000000
Depth: 997, time: 0:00:00.000000
Depth: 998, time: 0:00:00.000000
Depth: 999, time: 0:00:00.000000
```

Reply:
Interesting, I'm surprised.  I think `yield* TinyJson(element).leaves;` is equivalent to `for(var e in TinyJson(element).leaves) yield e`, and with the right input, I think `TinyJson(element).leaves` should be a stack of iterables of depth `n`, and so the `ith` `.next` call should have to traverse `i` stacked iterables to find the next element, hence quadratic.

@rakudrama am I getting myself confused here (I admit to finding sync* confusing) or is this one of the patterns that dart2js and the VM are now able to recognize and optimize?

Reply:
(I didn't see this comment before I landed, I'll make changes as needed in a new PR).

Reply:
FWIW, if I build lists of increasing length using:
```
dynamic mk(int i) {
  if (i == 0) {
    return [];
  }
  return [i, mk(i-1)];
}
```

then the VM seems to exhibit linear time growth, but DDC exhibits quadratic growth:

```
128 elements takes 3000 ms  (23.4375 ms/elements)
256 elements takes 11000 ms  (42.96875 ms/elements)
512 elements takes 51000 ms  (99.609375 ms/elements)
1024 elements takes 212000 ms  (207.03125 ms/elements)
2048 elements takes 889000 ms  (434.08203125 ms/elements)
4096 elements takes 4496000 ms  (1097.65625 ms/elements)
```

So looks like maybe this is quadratic, but the VM (and maybe dart2js?) are able to optimize it away?